File: second-year/second-semester/MPP/mpp-proiect-repository-raulbaciulescu/travelAgency/Persistence/src/main/java/impl/database/FlightTable.java
package impl.database;

import api.Constants;
import api.Table;
import domain.dto.FlightDto;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.util.*;

public class FlightTable implements Table<Long, FlightDto> {
    private final Connection connection;
    private final Map<Constants.Db.Queries, PreparedStatement> statements;
    private static final Logger logger = LogManager.getLogger();

    public FlightTable(Connection connection) throws SQLException {
        logger.info("Initializing FlightTable");
        this.statements = new HashMap<>();
        this.connection = connection;
        initStatements();
    }

    void initStatements() throws SQLException {
        statements.put(Constants.Db.Queries.ADD,
                connection.prepareStatement("INSERT INTO flight (id, startId, destinationId, startDate, nrOfSeats) VALUES (?, ?, ?, ?, ?)"
                ));
        statements.put(Constants.Db.Queries.DELETE,
                connection.prepareStatement("DELETE FROM flight WHERE id = ?"
                ));
        statements.put(Constants.Db.Queries.FIND_BY_ID,
                connection.prepareStatement("SELECT * FROM flight WHERE id = ?"
                ));
        statements.put(Constants.Db.Queries.GET_ALL,
                connection.prepareStatement("SELECT * FROM flight"
                ));
        statements.put(Constants.Db.Queries.UPDATE,
                connection.prepareStatement("UPDATE flight SET nrOfSeats = ?, startId = ?, destinationId = ?, startDate = ? WHERE id = ?"));
    }

    @Override
    public void add(FlightDto flightDto) {
        logger.traceEntry("saving flight {} ", flightDto);
        PreparedStatement statement = statements.get(Constants.Db.Queries.ADD);
        try {
            statement.setLong(1, flightDto.getId());
            statement.setLong(2, flightDto.getStartId());
            statement.setLong(3, flightDto.getDestinationId());
            statement.setString(4, flightDto.getStartDate().toString());
            statement.setInt(5, flightDto.getNrOfSeats());
            int result = statement.executeUpdate();
            logger.trace("Saved {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Long id) {
        logger.traceEntry("delete flight with id {} ", id);
        PreparedStatement statement = statements.get(Constants.Db.Queries.DELETE);
        try {
            statement.setLong(1, id);
            int result = statement.executeUpdate();
            logger.trace("Deleted {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }

    @Override
    public void update(FlightDto elem, FlightDto newElem) {
        logger.traceEntry();
        PreparedStatement statement = statements.get(Constants.Db.Queries.UPDATE);
        try {
            statement.setInt(1, newElem.getNrOfSeats());
            statement.setLong(2, newElem.getStartId());
            statement.setLong(3, newElem.getDestinationId());
            statement.setString(4, newElem.getStartDate().toString());
            statement.setLong(5, elem.getId());
            int result = statement.executeUpdate();
            logger.trace("Updated {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }

    @Override
    public Optional<FlightDto> findById(Long id) {
        logger.traceEntry();
        List<FlightDto> flights = new ArrayList<>();
        PreparedStatement statement = statements.get(Constants.Db.Queries.FIND_BY_ID);
        try {
            statement.setLong(1, id);
            try (ResultSet result = statement.executeQuery()) {
                while (result.next()) {
                    long id1 = result.getLong("id");
                    long startId = result.getLong("startId");
                    long destinationId = result.getLong("destinationId");
                    LocalDateTime startDate = LocalDateTime.parse(result.getString("startDate"));
                    int nrOfSeats = result.getInt("nrOfSeats");
                    FlightDto flightDto = new FlightDto(startId, destinationId, startDate, nrOfSeats);
                    flightDto.setId(id1);
                    flights.add(flightDto);
                }
            }
        } catch (SQLException throwables) {
            logger.error(throwables);
            System.out.println("Error db" + throwables);
        }
        logger.traceExit();
        if (flights.size() > 0)
            return Optional.of(flights.get(0));
        else
            return Optional.empty();
    }

    @Override
    public List<FlightDto> getAll() {
        logger.traceEntry();
        List<FlightDto> flights = new ArrayList<>();
        PreparedStatement statement = statements.get(Constants.Db.Queries.GET_ALL);
        try {
            try (ResultSet result = statement.executeQuery()) {
                while (result.next()) {
                    long id1 = result.getLong("id");
                    long startId = result.getLong("startId");
                    long destinationId = result.getLong("destinationId");
                    LocalDateTime startDate = LocalDateTime.parse(result.getString("startDate"));
                    int nrOfSeats = result.getInt("nrOfSeats");
                    FlightDto flightDto = new FlightDto(startId, destinationId, startDate, nrOfSeats);
                    flightDto.setId(id1);
                    flights.add(flightDto);
                }
            }
        } catch (SQLException throwables) {
            logger.error(throwables);
            System.out.println("Error db" + throwables);
        }
        logger.traceExit();
        return flights;
    }
}

File: second-year/second-semester/MPP/mpp-proiect-repository-raulbaciulescu/travelAgency2/persistance/database/FlightTable.cs
using System.Data.SQLite;
using Microsoft.VisualBasic;
using travelAgency2.Repository.Database;

using model;
using model.Dto;
using model.Utils;

namespace persistance.Database
{
    internal class FlightTable : ITable<long, FlightDto>
    {
        private SQLiteConnection connection;
        private Dictionary<Db.Queries, SQLiteCommand> statements;
        private static readonly log4net.ILog logger = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

        public FlightTable(SQLiteConnection connection)
        {
            logger.Info("Initializing FlightTable");
            this.connection = connection;
            this.statements = new Dictionary<Db.Queries, SQLiteCommand>();
            InitStatements();
        }

        private void InitStatements()
        {

            SQLiteCommand command1 = new SQLiteCommand(connection);
            SQLiteCommand command2 = new SQLiteCommand(connection);
            SQLiteCommand command3 = new SQLiteCommand(connection);
            SQLiteCommand command4 = new SQLiteCommand(connection);
            SQLiteCommand command5 = new SQLiteCommand(connection);
            command1.CommandText = "INSERT INTO flight (id, startId, destinationId, startDate, nrOfSeats) " +
                "VALUES (@id, @startId, @destinationId, @startDate, @nrOfSeats)";
            command2.CommandText = "DELETE * FROM flight WHERE id = @id";
            command3.CommandText = "SELECT * FROM flight WHERE id = @id";
            command4.CommandText = "SELECT * FROM flight";
            command5.CommandText = "UPDATE flight SET nrOfSeats = @nrOfSeats WHERE id = @id";
            statements.Add(Db.Queries.ADD, command1);
            statements.Add(Db.Queries.DELETE, command2);
            statements.Add(Db.Queries.FIND_BY_ID, command3);
            statements.Add(Db.Queries.GET_ALL, command4);
            statements.Add(Db.Queries.UPDATE, command5);
        }

        public void Add(FlightDto flight)
        {
            logger.Info("enter in add flight");
            SQLiteCommand statement = statements[Db.Queries.ADD];
            try
            {
                statement.Parameters.AddWithValue("@id", flight.Id);
                statement.Parameters.AddWithValue("@startId", flight.startId);
                statement.Parameters.AddWithValue("@destinationId", flight.destinationId);
                statement.Parameters.AddWithValue("@startDate", flight.startDate.ToString());
                statement.Parameters.AddWithValue("@nrOfSeats", flight.nrOfSeats);
                int numberofrowsaffected = statement.ExecuteNonQuery();
                logger.InfoFormat("Saved {0} instances", numberofrowsaffected);
            }
            catch (Exception ex)
            {
                logger.Error(ex);
                Console.WriteLine(ex.Message);
            }
            logger.Info("exit from add flight");
        }
        public void Delete(long id)
        {
            //TODO
        }

        public void Update(FlightDto flightDto, FlightDto flightDto1)
        {
            logger.Info("enter in update flight");
            SQLiteCommand statement = statements[Db.Queries.UPDATE];
            try
            {
                statement.Parameters.AddWithValue("@id", flightDto.Id);
                statement.Parameters.AddWithValue("@nrOfSeats", flightDto1.nrOfSeats);
                int numberofrowsaffected = statement.ExecuteNonQuery();
                logger.InfoFormat("Updated {0} instances", numberofrowsaffected);
            }
            catch (Exception ex)
            {
                logger.Error(ex);
                Console.WriteLine(ex.Message);
            }
            logger.Info("exit from update flight");
        }
        public FlightDto FindById(long id)
        {
            logger.Info("enter in findById flight");
            List<FlightDto> flights = new List<FlightDto>();
            SQLiteCommand statement = statements[Db.Queries.FIND_BY_ID];
            statement.Parameters.AddWithValue("@id", id);
            SQLiteDataReader reader = statement.ExecuteReader();
            if (reader.HasRows)
            {
                while (reader.Read())
                {
                    FlightDto flightDto = new FlightDto((long)reader.GetInt64(4),
                        (long)reader.GetInt64(0), (long)reader.GetInt64(1), DateTime.Parse(reader.GetString(2)),
                        reader.GetInt32(3));
                    flights.Add(flightDto);
                }
            }
            reader.Close();
            logger.Info("exit from findById location");
            return flights[0];
        }

        public List<FlightDto> GetAll()
        {
            logger.Info("enter in getAll flight");
            List<FlightDto> flights = new List<FlightDto>();
            SQLiteCommand statement = statements[Db.Queries.GET_ALL];
            SQLiteDataReader reader = statement.ExecuteReader();
            if (reader.HasRows)
            {
                while (reader.Read())
                {
                    FlightDto flightDto = new FlightDto((long)reader.GetInt64(4),
                        (long)reader.GetInt64(0), (long)reader.GetInt64(1), DateTime.Parse(reader.GetString(2)),
                        reader.GetInt32(3));
                    flights.Add(flightDto);
                }
            }
            reader.Close();
            logger.Info("exit from getAll location");
            return flights;
        }
    }
}
